jetty-http/src/test/java/org/eclipse/jetty/http/HttpURITest.java

```diff
@@ -401,10 +421,47 @@ public void testDecodedPath(String input, String decodedPath, EnumSet<Ambiguous>
             assertThat(uri.hasAmbiguousSegment(), is(expected.contains(Ambiguous.SEGMENT)));
             assertThat(uri.hasAmbiguousSeparator(), is(expected.contains(Ambiguous.SEPARATOR)));
             assertThat(uri.hasAmbiguousParameter(), is(expected.contains(Ambiguous.PARAM)));
+            assertThat(uri.hasAmbiguousEncoding(), is(expected.contains(Ambiguous.ENCODING)));
         }
         catch (Exception e)
         {
             assertThat(decodedPath, nullValue());
         }
     }
+
+    public static Stream<Arguments> emptySegmentTests()
+    {
+        return Arrays.stream(new Object[][]
+            {
+                // Empty segment tests.
+                {"/", EnumSet.noneOf(Ambiguous.class)},
+                {"/#", EnumSet.noneOf(Ambiguous.class)},
+                {"/path", EnumSet.noneOf(Ambiguous.class)},
+                {"/path/", EnumSet.noneOf(Ambiguous.class)},
+                {"//", EnumSet.of(Ambiguous.EMPTY)},
+                {"/foo//", EnumSet.of(Ambiguous.EMPTY)},
+                {"/foo//bar", EnumSet.of(Ambiguous.EMPTY)},
+                {"//foo/bar", EnumSet.of(Ambiguous.EMPTY)},
+                {"/foo?bar", EnumSet.noneOf(Ambiguous.class)},
+                {"/foo#bar", EnumSet.noneOf(Ambiguous.class)},
+                {"/foo;bar", EnumSet.noneOf(Ambiguous.class)},
+                {"/foo/?bar", EnumSet.noneOf(Ambiguous.class)},
+                {"/foo/#bar", EnumSet.noneOf(Ambiguous.class)},
+                {"/foo/;param", EnumSet.noneOf(Ambiguous.class)},
+                {"/foo/;param/bar", EnumSet.of(Ambiguous.EMPTY)},
+            }).map(Arguments::of);
+    }
+
+    @ParameterizedTest
+    @MethodSource("emptySegmentTests")
+    public void testEmptySegment(String input, EnumSet<Ambiguous> expected)
```

Comment:
Can't these tests all be done in the test above?

Reply:
No because for this test we are using `HttpURI.from("GET", input)` instead of just `HttpURI.from(input)`, this starts us from state `PATH` so it can reproduce some failures the other one cannot.

There are tests for empty segments in the above one as well.

Reply:
Then rename the test to `testFromPathQuery` and use `HttpURI.build().pathQuery(uri)` instead of the `from` method.    You can then add a bunch of other tests (eg %2f and %2e ones) that verify expected behaviour when parsing starting from a path or query